### PR TITLE
Adds config transformation template for RDS source

### PR DESCRIPTION
**Pull request by oeyh**
### Description
This PR adds config transformation template for RDS source:
- Slightly updates RDS source config: moves partition_count to top level; adds role arn for export config
- For the S3 bucket, on top of user provided s3 prefix, add source coordinator prefix as subpath to distinguish between different pipelines. Under the subpath, further splits into two prefixes: `rds-export` for RDS exported data and `buffer` for pipeline generated data. 
- Refactors `ExportRecordConverter` and `StreamRecordConverter` to extract common code
- Adds rule and template for config transformation. The template is mostly adopted from documentdb ([link](https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/mongodb/src/main/resources/org/opensearch/dataprepper/transforms/templates/documentdb-template.yaml)), with changes to the functions to get `include_prefix` and `depth`.

### Testing
Tested a pipeline with RDS source and verified the config transformation works and the pipeline processes export and stream data as expected.

 
### Issues Resolved
Contributes to #4561 
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by chenqi0805 — APPROVED** [review]

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/RdsSourceConfig.java [hunk 1]
```diff
@@ -21,6 +24,7 @@
  * Configuration for RDS Source
  */
 public class RdsSourceConfig {
+    private static final int DEFAULT_S3_FOLDER_PARTITION_COUNT = 100;
```
Comment (dinujoh) [inline data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/RdsSourceConfig.java:27]: Is the same default value for both export and stream ?

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/model/ExportObjectKey.java [hunk 2]
```diff
@@ -29,18 +33,21 @@ public class ExportObjectKey {
 
     public static ExportObjectKey fromString(final String objectKeyString) {
 
-        final String[] parts = objectKeyString.split("/");
-        if (parts.length != 6) {
+        final String[] parts = objectKeyString.split(S3_PATH_DELIMITER);
+        if (parts.length < 5) {
```
Comment (dinujoh) [inline data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/model/ExportObjectKey.java:37]: any reason this change ?

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java [hunk 3]
```diff
@@ -244,16 +246,20 @@ private void handleRowChangeEvent(com.github.shyiko.mysql.binlog.event.Event eve
                 rowDataMap.put(columnNames.get(i), rowDataArray[i]);
             }
 
+            final Event dataPrepperEvent = JacksonEvent.builder()
+                    .withEventType("event")
```
Comment (dinujoh) [inline data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java:250]: move this to constant.

File: data-prepper-plugins/rds-source/src/test/java/org/opensearch/dataprepper/plugins/source/rds/leader/LeaderSchedulerTest.java [hunk 4]
```diff
@@ -138,6 +140,6 @@ void test_shutDown() {
     }
 
     private LeaderScheduler createObjectUnderTest() {
-        return new LeaderScheduler(sourceCoordinator, sourceConfig, schemaManager, dbMetadata);
+        return new LeaderScheduler(sourceCoordinator, sourceConfig, s3Prefix, schemaManager, dbMetadata);
```
Comment (dinujoh) [inline data-prepper-plugins/rds-source/src/test/java/org/opensearch/dataprepper/plugins/source/rds/leader/LeaderSchedulerTest.java:143]: Are we storing S3 partition info in Source coordinator ? If yes, we can try to assert S3 prefix value in S3 Partition.

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/DynamicConfigTransformer.java [hunk 5]
```diff
@@ -402,15 +403,28 @@ private boolean isJsonPath(String parameter) {
...
+        return Integer.toString(getDepth(s3Prefix, 4));
+    }
+
+    /**
+     * Calculate s3 folder scan depth for RDS source pipeline
+     * @param s3Prefix: s3 prefix defined in the source configuration
+     * @return s3 folder scan depth
+     */
+    public String calculateDepthForRdsSource(String s3Prefix) {
+        return Integer.toString(getDepth(s3Prefix, 3));
```
Comment (dinujoh) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/DynamicConfigTransformer.java:420]: Why is the difference between RDS and DocDB ?

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/RdsSourceConfig.java (hunk 1 above)
Comment (oeyh) [reply data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/RdsSourceConfig.java:27]: Yes

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/model/ExportObjectKey.java (hunk 2 above)
Comment (oeyh) [reply data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/model/ExportObjectKey.java:37]: The previous code doesn't take into account where prefix has multiple levels (eg, "outer/inner") or is empty (eg, "").

File: data-prepper-plugins/rds-source/src/test/java/org/opensearch/dataprepper/plugins/source/rds/leader/LeaderSchedulerTest.java (hunk 4 above)
Comment (oeyh) [reply data-prepper-plugins/rds-source/src/test/java/org/opensearch/dataprepper/plugins/source/rds/leader/LeaderSchedulerTest.java:143]: S3 partition info is not stored in source coordinator.

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/DynamicConfigTransformer.java (hunk 5 above)
Comment (oeyh) [reply data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/DynamicConfigTransformer.java:420]: I removed the timestamp in EpochMillis from S3 path, so it is one layer shorter compared to DocumentDB.

Comment (oeyh) [reply data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/DynamicConfigTransformer.java:420]: Discussed offline. Will make it consistent with DocDB.

Comment (oeyh) [reply data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/DynamicConfigTransformer.java:420]: Discussed again. The S3 prefix looks like this: `user defined prefix / source identifier / buffer / folder partitions`. So I have depth = length of user defined prefix + 3.

File: data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java (hunk 3 above)
Comment (oeyh) [reply data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java]: Done.

File: data-prepper-plugins/rds-source/src/test/java/org/opensearch/dataprepper/plugins/source/rds/leader/LeaderSchedulerTest.java (hunk 4 above)
Comment (oeyh) [reply data-prepper-plugins/rds-source/src/test/java/org/opensearch/dataprepper/plugins/source/rds/leader/LeaderSchedulerTest.java:143]: Added a unit test to assert the s3 prefix

**Review by dinujoh — APPROVED** [review]

**Review by chenqi0805 — APPROVED** [review]

**Pull request merged**